### PR TITLE
RabbitMQ and OSLO: Start RPC server/client and modify IPTables

### PR DESCRIPTION
**Pull request by alex7c4**
QA-2295 - Start RabbitMQ RPC server and client and apply IPTABLES DROP rules for RabbitMQ ports on one controller
QA-2296 - Start RabbitMQ RPC server and client and apply IPTABLES REJECT rules for RabbitMQ ports on one controller

### CODE REVIEW

File: mos_tests/rabbitmq_oslo/conftest.py [hunk 1]
```diff
@@ -38,3 +44,57 @@ def fixt_kill_rpc_server_client(env):
...
+
+
+@pytest.yield_fixture
+def fixt_iptables_drop_reject(env, request):
+    """Choose one controller and apply IPTables rules"""
+    i_am_chosen_mark = '/tmp/i_am_chosen_mark.txt'
+
+    if request.param == 'drop':
+        table_modif = (
+            'iptables -I INPUT -p tcp -m tcp --dport 4369  -j DROP ; '
```
Comment (gdyuldin) [inline mos_tests/rabbitmq_oslo/conftest.py]: Let's make 2 command templates like `iptables -{action} INPUT -p tcp -m tcp --dport {port}  -j DROP`
And do smth like `remote.execute(cmd_template.format(action='I', port=4369))` (perhaps in loop over ports)

File: mos_tests/rabbitmq_oslo/conftest.py [hunk 2]
```diff
@@ -38,3 +44,57 @@ def fixt_kill_rpc_server_client(env):
...
+        raise ValueError("Does'n know such param [{0}]!".format(request.param))
+
+    controller = random.choice(env.get_nodes_by_role('controller'))
+    logger.debug('Applying IPTables rules [{0}] to {1}'.format(
+        request.param, controller.data['ip']))
+    with controller.ssh() as remote:
+        # execute iptables DROP command
+        remote.check_call(table_modif)
+        # mark this controller
+        with remote.open(i_am_chosen_mark, 'w') as f:
```
Comment (gdyuldin) [inline mos_tests/rabbitmq_oslo/conftest.py]: If you need to store, on which controller your executed commands, more py.test way is to determine fixture - `controller`.

File: mos_tests/rabbitmq_oslo/conftest.py [hunk 3]
```diff
@@ -38,3 +44,57 @@ def fixt_kill_rpc_server_client(env):
         with node.ssh() as remote:
             cmd = 'pkill -f oslo_msg_check_'
             remote.execute(cmd)
+
+
+@pytest.yield_fixture
+def fixt_iptables_drop_reject(env, request):
```
Comment (gdyuldin) [inline mos_tests/rabbitmq_oslo/conftest.py]: Let's call it like `patch_iptables` (or smth more suitable to real action of fixture)

File: mos_tests/rabbitmq_oslo/conftest.py [hunk 4]
```diff
@@ -38,3 +44,57 @@ def fixt_kill_rpc_server_client(env):
...
+            ' -j REJECT --reject-with icmp-host-prohibited ; ')
+        table_modif_del = (
+            'iptables -D INPUT -p tcp -m tcp --dport 4369'
+            ' -j REJECT --reject-with icmp-host-prohibited ; '
+            'iptables -D INPUT -p tcp -m tcp --dport 25672'
+            ' -j REJECT --reject-with icmp-host-prohibited ; '
+            'iptables -D INPUT -p tcp -m tcp --dport 5672'
+            ' -j REJECT --reject-with icmp-host-prohibited ; '
+            'iptables -D INPUT -p tcp -m tcp --dport 5673'
+            ' -j REJECT --reject-with icmp-host-prohibited ; ')
```
Comment (schipiga) [inline mos_tests/rabbitmq_oslo/conftest.py]: May be?

``` python
def add_ports(tmpl):
    return ' '.join(map(lambda port: tmpl.format(port),
                        (4369, 25672, 5673, 5673)))

if request.param == 'drop':
    table_modif = add_ports(
        'iptables -I INPUT -p tcp -m tcp --dport {} -j DROP ;')
    table_modif_del = add_ports(
        'iptables -D INPUT -p tcp -m tcp --dport {} -j DROP ;')
elif request.param == 'reject':
    table_modif = add_ports('iptables -I INPUT -p tcp -m tcp --dport {}'
                            ' -j REJECT --reject-with icmp-host-prohibited ;')
    table_modif_del = add_ports('iptables -D INPUT -p tcp -m tcp --dport {} -j'
                                ' REJECT --reject-with icmp-host-prohibited ;')
```

File: mos_tests/rabbitmq_oslo/conftest.py [hunk 5]
```diff
@@ -38,3 +44,57 @@ def fixt_kill_rpc_server_client(env):
         with node.ssh() as remote:
             cmd = 'pkill -f oslo_msg_check_'
             remote.execute(cmd)
+
+
+@pytest.yield_fixture
+def fixt_iptables_drop_reject(env, request):
+    """Choose one controller and apply IPTables rules"""
+    i_am_chosen_mark = '/tmp/i_am_chosen_mark.txt'
```
Comment (schipiga) [inline mos_tests/rabbitmq_oslo/conftest.py]: maybe use:

``` python
import tempfile

_, file_name = tempfile.mkstemp()
```

File: mos_tests/rabbitmq_oslo/conftest.py (hunk 3 above)
Comment (schipiga) [inline mos_tests/rabbitmq_oslo/conftest.py]: agree, prefixes look ugly :)

File: mos_tests/rabbitmq_oslo/conftest.py (hunk 2 above)
Comment (schipiga) [inline mos_tests/rabbitmq_oslo/conftest.py]: :+1:

File: mos_tests/rabbitmq_oslo/conftest.py [hunk 6]
```diff
@@ -38,3 +44,43 @@ def fixt_kill_rpc_server_client(env):
...
+    return random.choice(env.get_nodes_by_role('controller'))
+
+
+@pytest.yield_fixture
+def patch_iptables(controller, request):
+    """Apply IPTables rules"""
+
+    def add_ports(tmpl):
+        ports = [4369, 5672, 5673, 25672]
+        return ' '.join(map(lambda portnum: tmpl.format(portnum=portnum),
```
Comment (gdyuldin) [inline mos_tests/rabbitmq_oslo/conftest.py]: `' '.join([tmpl.format(portnum=x) for x in ports])`

File: mos_tests/rabbitmq_oslo/conftest.py [hunk 7]
```diff
@@ -38,3 +44,43 @@ def fixt_kill_rpc_server_client(env):
...
+                            ports))
+
+    if request.param == 'drop':
```
Comment (gdyuldin) [inline mos_tests/rabbitmq_oslo/conftest.py]: Let's use 'DROP' and 'REJECT' as request.param and templates will looks like
'iptables -I INPUT -p tcp -m tcp --dport {portnum} -j {action} ;',
where action will pass directly from request.param

File: mos_tests/rabbitmq_oslo/conftest.py (hunk 6 above)
Comment (schipiga) [inline mos_tests/rabbitmq_oslo/conftest.py]: @gdyuldin that's cool :+1:

File: mos_tests/rabbitmq_oslo/conftest.py [hunk 8]
```diff
@@ -38,3 +44,43 @@ def fixt_kill_rpc_server_client(env):
...
+        tbl_modif_del = add_ports(
+            'iptables -D INPUT -p tcp -m tcp --dport {portnum} -j REJECT'
+            ' --reject-with icmp-host-prohibited ;')
+    else:
+        raise ValueError("Does'n know such param [{0}]!".format(request.param))
+
+    logger.debug('Applying IPTables rules [{0}] to {1}'.format(
+        request.param, controller.data['ip']))
+    with controller.ssh() as remote:
+        remote.check_call(tbl_modif)
```
Comment (gdyuldin) [inline mos_tests/rabbitmq_oslo/conftest.py]: Command to execute contains ';' as delimiter. This mean, that only if last command returns not zero exit code - exception will generate. It's better to join commands with `&&` or call each separately, like:

``` python
with controller.ssh() as remote:
    for cmd in cmds:
        remote.check_call(cmd)
```

Comment (gdyuldin) [inline mos_tests/rabbitmq_oslo/conftest.py]: In last case you can iterate over ports.

**Pull request merged**